go-controller/pkg/ovn/egressfirewall.go

```diff
@@ -165,59 +170,149 @@ func (oc *Controller) updateEgressFirewallWithRetry(egressfirewall *egressfirewa
 	})
 }
 
-func (ef *egressFirewall) addACLToJoinSwitch(joinSwitches []string, hashedAddressSetNameIPv4, hashedAddressSetNameIPv6 string) error {
+func (oc *Controller) applyEgressFirewall(joinSwitches []string, hashedAddressSetNameIPv4, hashedAddressSetNameIPv6 string, ef *egressFirewall) error {
 	for _, rule := range ef.egressRules {
-		var match string
 		var action string
+		var matchTargets []matchTarget
 		if rule.access == egressfirewallapi.EgressFirewallRuleAllow {
 			action = "allow"
 		} else {
 			action = "drop"
 		}
+		if rule.to.cidrSelector != "" {
+			if utilnet.IsIPv6CIDRString(rule.to.cidrSelector) {
+				matchTargets = []matchTarget{{matchKindV6CIDR, rule.to.cidrSelector}}
+			} else {
+				matchTargets = []matchTarget{{matchKindV4CIDR, rule.to.cidrSelector}}
 
-		ipAddress, _, err := net.ParseCIDR(rule.to.cidrSelector)
-		if err != nil {
-			return fmt.Errorf("error rule.to.cidrSelector %s is not a valid CIDR (%+v)", rule.to.cidrSelector, err)
-		}
-		if !utilnet.IsIPv6(ipAddress) {
-			match = fmt.Sprintf("match=\"ip4.dst == %s && ip4.src == $%s\"", rule.to.cidrSelector, hashedAddressSetNameIPv4)
+			}
 		} else {
-			match = fmt.Sprintf("match=\"ip6.dst == %s && ip6.src == $%s\"", rule.to.cidrSelector, hashedAddressSetNameIPv6)
+			// rule based on DNS NAME
+			if rule.access == egressfirewallapi.EgressFirewallRuleDeny {
+				return fmt.Errorf("do not use dnsName as part of a deny rule this is insecure")
+			}
+			dnsNameAddressSets, err := oc.egressFirewallDNS.Add(ef.namespace, rule.to.dnsName)
+			if err != nil {
+				return fmt.Errorf("error with EgressFirewallDNS - %v", err)
+			}
+			if dnsNameAddressSets.GetIPv4HashName() != "" {
+				matchTargets = append(matchTargets, matchTarget{matchKindV4AddressSet, dnsNameAddressSets.GetIPv4HashName()})
+			}
+			if dnsNameAddressSets.GetIPv6HashName() != "" {
+				matchTargets = append(matchTargets, matchTarget{matchKindV6AddressSet, dnsNameAddressSets.GetIPv6HashName()})
+			}
 		}
-
-		if len(rule.ports) > 0 {
-			match = fmt.Sprintf("%s && ( %s )\"", match[:len(match)-1], egressGetL4Match(rule.ports))
+		match := generateMatch(hashedAddressSetNameIPv4, hashedAddressSetNameIPv6, matchTargets, rule.ports)
+		for _, joinSwitch := range joinSwitches {
+			err := createACLRule(defaultStartPriority-rule.id, match, action, ef.namespace, joinSwitch)
```

Comment:
Another thought: what happens when a new Node is created afterwards? Will it be missing all the egress firewalls?

Can you think of a better architecture? Maybe a portgroup that has every join switch, then add and remove acls to that portgroup?

Reply:
It will be missing the egressFirewalls, a check for egressFirewalls  in the `WatchNodes()` `AddFunc()` https://github.com/ovn-org/ovn-kubernetes/blob/c10672377bd829ceee69c6ab90cc1ee0c6940f18/go-controller/pkg/ovn/ovn.go#L971-L974
this checks for existing egressfirewalls in every namespace and adds the ACL to the new nodes joinSwitch if one exists